### ISSUE TITLE
Escaping illegal characters

### ISSUE THREAD

**Issue by InteXX** [VBAndCs/Vazor#15]
Did you figure out how to get past the problem of illegal characters in XML Literals? For example, `&nbsp;` and `&mdash;` are commonly used in HTML markup, but a project won't compile if we use them in Vazor. The problem is the standalone ampersand (`&`).

The best workaround I could find is to use `&amp;` in the literals and then replace them all in `*.cshtml.vb`:
```
html = Me.GetVbXml(Me.ViewData).ParseZML.Replace("&amp;", "&")
```
But of course that's a kludge.

Any ideas for an elegant solution to this?

**Comment by VBAndCs** [comment]
Try something like this:
```
        Const nbsp = "&nbsp"
        Dim x = <html>
                    <b><%= nbsp %></b>
                </html>
```
________________________________
From: Jeff Bowman ***@***.***>
Sent: Sunday, November 12, 2023 4:23 AM
To: VBAndCs/Vazor ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [VBAndCs/Vazor] Escaping illegal characters (Issue #15)


Did you figure out how to get past the problem of illegal characters in XML Literals? For example, &nbsp; and &mdash; are commonly used in HTML markup, but a project won't compile if we use them in Vazor. The problem is the standalone ampersand (&).

The best workaround I could find is to use &amp; in the literals and then replace them all in *.cshtml.vb:

html = Me.GetVbXml(Me.ViewData).ParseZML.Replace("&amp;", "&")


But of course that's a kludge.

Any ideas for an elegant solution to this?

**Comment by InteXX** [comment]
Got it, thanks.

**Comment by InteXX** [comment]
Oops, I spoke too soon.

Turns out that inserts an ampersand entity into the HTML (`&amp;`), which mucks things up.

The end result: `&amp;nbsp;`. Which of course displays a `&nbsp;` on the page, where there should be a non-breaking space instead.

Guess we'll have to go with the kludge.

**Comment by InteXX** [comment]
OK, I found the fix. Use the HTML entity `&#38;` instead of a bare ampersand (`&`).

For example: `&#38;nbsp;`

Now we don't have to muck around with replacing stuff in `*.cshtml.vb`.

**Comment by InteXX** [comment]
Damn... that one doesn't work either! No matter what we try, the XML Literals engine escapes that ampersand.

It seems the only way to do it is to modify the text _after_ the engine generates it, not before. Runtime only. We don't get design-time support on this one.

And the engine replaces `&#38;` with `&amp;`, so we might as well use `&amp;` in the first place. Easier to type, easier to remember.

**Comment by VBAndCs** [comment]
In ZML source code, I use the const `__amp__`  which is not likely to be found in the original page, then replace it with `&` in the final text.

________________________________

**Comment by InteXX** [comment]
OK.

In any case, the only solution seems to be a combination designtime/runtime solution. If the author wants to use HTML entities such as `&nbsp;` and `&mdash;`, he must use special notations (designtime) and then modify ZML's output after the fact (runtime).

Unfortunately, this results in brittle code. But so far as I can tell, it's the only answer to a sticky problem.

**Comment by VBAndCs** [comment]
In ZML, you can find this note:
>

You can use >, < and & symbols directly in ZML code, as it handles them internally. Don't encode these symbols unless you intend to view them in the browser. But note that using these symbols in vbxml literals is not valid, unless you embed the string containing them in  <%= %> quotes to force VB to encode them for you.

So, if you use ZML files, you need to do nothing, but if you write XML literals in VB code, you just need to define a place holder and add an extra line to replace it in the output. I think leaving this to the coder is better, instead of forcing him to memories some special placeholders for & and alike.

________________________________

**Comment by InteXX** [comment]
Makes sense.

I may be getting ZML and XML Literals mixed up in my terminology.

I'm going by the sample code in `Index.vbhtml.vb`:
```
<zml xmlns:z="zml">
   ... HTML markup here ...
</zml>
```
I've been thinking that everything here is encompassed by the concept of XML Literals, and that ZML is everything between `<zml>` and `</zml>`.

Is that correct?

**Comment by VBAndCs** [comment]
No. This is still xml literal, that contains html tags and possibly some zml tags to fix some issues in for loops. On the other hand there are pure .zml files, where you can design the page with html and zml tag language. See:
https://github.com/VBAndCs/ZML
The missing part in this project is thee editor support for zml tags.
Note: I no longer have access to my github account, because I don't use 2 factor authintication, and github doesn't support phone messages to Egypt! I will not install any extension to the browser not use any mobile app to scan codes. In fact github was sending the authentication code via email, and that was a 2 factor already, but now it just complicated things.
So, I can just replay on your messages via email. There will be no further activity on my projects until github supports SMS or go back to the email auth.
So, it is better to fork such projects and evolve it as you want.
Thanks.

________________________________

**Comment by InteXX** [comment]
> I no longer have access to my github account

What!!

That'll never do! OK, I'm going to help you get back in there.

First, get a Google Voice number in the U.S. (free). Then, send me an email so we can communicate that way (much easier). My email address is on my website.
